### PR TITLE
Prevent rendering of hidden windows.

### PR DESCRIPTION
**Pull request by Rodbourn**
This is a cleaner pull request for #3985 (deleted the fork to get one free of the merge).

This simply prevents rendering of windows with a viewport that are covered by another window (at least covering it, they can be bigger).

### CODE REVIEW

**Comment by janisozaur** [comment]
There is no need to open another PR just because you want to rebase. Simply replace commits in your branch and that's all.

But it doesn't matter any more, as it is done, I'll close the other one.

**Comment by mrtnptrs** [comment]
@janisozaur Try to sound a bit less hostile, we all know you mean it good, but.... it sounds a bit like you hate the one you speak to :)

@Rodbourn the following issue still occurs when I download the artifact/build from this pull request: _this all works perfectly and the FPS stays at a reasonable level (around 42 FPS) when you don't move around the map with many staff windows opened over each other. But as soon as I move around the map the FPS get a big hit again: from 10 to 20 FPS (does not occur when moving around the map without all the window open) It seems like the windows and viewports don't get rederd when you are not moving around the map, but when you move around the map they DO get rendered. (I'm not sure of this as I'm not a programmer!)_ 

So, that issue is not fixed yet.

File: src/interface/window.c [hunk 1]
```diff
@@ -2477,3 +2479,29 @@ void window_update_textbox()
...
+		if (w_other->x <= w->x
+			&& w_other->y <= w->y
+			&& w_other->x + w_other->width >= w->x + w->width
+			&& w_other->y + w_other->height >= w->y + w->height)
+		{
+			return false;
+		}
+	}
+
+	// default to previous behavior
```
Comment (janisozaur) [inline src/interface/window.c]: behavio*u*r

Comment (Rodbourn) [inline src/interface/window.c]: Enforcing British English on comments? ;)

Comment (SijmenSchoon) [inline src/interface/window.c]: Considering Sawyer himself was a British bloke it'd only be a tribute to him to continue his legacy in the comments 😃

**Comment by Rodbourn** [comment]
@mrtnptrs I have tracked this issue down to the viewport invalidation which is triggered by scrolling.  I made a moderately ambitious update that will cause the viewports to skip invalidation if they are covered as well.  This method is hit pretty heavily (per-peep!), so I put in a 'cache' of sorts which will make repeat hits within a game cycle very cheap.

I tested this with a park containing 4,500 guests and 50 staff windows open with a profiler attached.  ~20 fps no windows, ~8 fps 50 windows, ~3-4 fps scrolling like crazy.  During the scrolling the viewport invalidation was doing 23% of the work, with the extra work to skip hidden viewports taking ~2% of the work.  This should scale up heavily without a profiler attached, of course.

File: src/drawing/engines/SoftwareDrawingEngine.cpp [hunk 2]
```diff
@@ -365,6 +365,8 @@ class SoftwareDrawingEngine : public IDrawingEngine
             _rainDrawer.SetDPI(&_bitsDPI);
             _rainDrawer.Restore();
 
+			ResetWindowVisbilities();
```
Comment (IntelOrca) [inline src/drawing/engines/SoftwareDrawingEngine.cpp]: Use spaces for cpp files.

File: src/interface/window.h [hunk 3]
```diff
@@ -493,6 +495,13 @@ enum {
 	MODAL_RESULT_OK
 };
 
+enum VISIBILITY_CACHE
+{
+	VC_UNKNOWN = 0,
+	VC_VISIBLE = 1,
+	VC_COVERED = 2
```
Comment (IntelOrca) [inline src/interface/window.h]: You don't need to define numbers for this enum.

**Comment by mrtnptrs** [comment]
@Rodbourn sounds good, but I'm not that good with understanding the coding-part, sorry :)

But does this pull request also work with the new OpenGL engine?

**Comment by Rodbourn** [comment]
@mrtnptrs so those numbers were with the visual studio profiler attached (instrumentation slows it down).

Some numbers from my end:

4,500 guests, software rendering (second option):

- no windows, stationary, ~66 fps
- no windows, scrolling around, ~50 fps
- 50 windows, stationary, ~66 fps
- 50 windows, scrolling around, ~38 fps

- OpenGL seems a bit odd in it's measurement, I get 125 FPS with windows scrolling
- Software

5960X @4.3Ghz, 64GB DDR4, 980 GTX, Win10

Anyway, it seems much better now than before.

**Comment by IntelOrca** [comment]
@Rodbourn OpenGL always draws everything every frame. But I expect your code improves the performance of that even more if it doesn't draw windows that can't be seen.

**Comment by mrtnptrs** [comment]
@Rodbourn This pull request really gives the game a boost with many windows opened. Does this also work when you open a few staff viewport windows next to each other, then open another windows (like the finances window) and let that window cover the staff viewport windows? And does it now only prevent the hidden viewports to be drawn or the whole window (with viewport)? That's not fully clear to me.

**Comment by Rodbourn** [comment]
@mrtnptrs if any window completely covers a window with a viewport it should help (such as the finance window over staff).  It won't handle the case that a window is covered by a combination of other windows, though.  Glad to hear it's helping!

**Comment by Nubbie** [comment]
Change log?

File: src/interface/window.h [hunk 4]
```diff
@@ -85,6 +85,7 @@ typedef struct rct_viewport {
 	uint8 zoom;						// 0x10
 	uint8 var_11;
 	uint16 flags;					// 0x12
+	uint8 visibility;				// VISIBILITY_CACHE
```
Comment (duncanspumpkin) [inline src/interface/window.h]: Update the viewport struct size comment if you are going to increase its size.

File: src/interface/window.h [hunk 5]
```diff
@@ -288,6 +289,7 @@ typedef struct rct_window {
 	sint8 var_4B8;
 	sint8 var_4B9;
 	uint8 colours[6];			// 0x4BA
+	uint8 visibility;			// VISIBILITY_CACHE
```
Comment (duncanspumpkin) [inline src/interface/window.h]: Update the the window struct size comment if you are going to increase its size.

File: src/interface/window.h (hunk 4 above)
Comment (IntelOrca) [inline src/interface/window.h]: It just needs removing now, its no longer vanillaised.

**Comment by Rodbourn** [comment]
@IntelOrca @duncanspumpkin I have the comment updates and a vanity update here https://github.com/Rodbourn/OpenRCT2, how shall I proceed to get those changes here?

**Comment by IntelOrca** [comment]
@Rodbourn you have to open a new PR

**Comment by Rodbourn** [comment]
@IntelOrca I'll do that, just wanted to be sure since I didn't have to last time ;)

**Pull request merged**